libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1000 @@
+
+#define EOSIO_CDT_CREATE_KV_INDICES(value_class, ...)                                                                  \
+   BOOST_PP_SEQ_FOR_EACH_I(EOSIO_CDT_CREATE_KV_INDEX, value_class, BOOST_PP_VARIADIC_TO_SEQ(__VA_ARGS__))
+
+#define EOSIO_CDT_LIST_KV_INDEX(r, data, index_name)                                                                   \
+   ,&index_name
+
+#define EOSIO_CDT_LIST_KV_INDICES(...)                                                                                 \
+   BOOST_PP_SEQ_FOR_EACH(EOSIO_CDT_LIST_KV_INDEX, _, BOOST_PP_VARIADIC_TO_SEQ(__VA_ARGS__))
+
+/**
+ * @brief Macro to define a table.
+ * @details The resulting table will have member fields on it that match 1-1 with the names of the
+ * fields passed into the list. See example for further clarification.
+ *
+ * @param table_class - The name of the class of the user defined table that inherits from eosio::kv_table
+ * @param value_class - The name of the class of the data stored as the value of the table
+ * @param table_name  - The name of the table
+ * @param db_name     - The type of the EOSIO Key Value database. Defaulted to eosio.kvram
+ * @param ...         - A variadic list of 1 or more indexes to define on the table.
+ */
+#define DEFINE_TABLE(table_class, value_class, table_name, db_name, /*indices*/...)                                    \
+   struct table_class : eosio::kv_table<value_class> {                                                                 \
+      EOSIO_CDT_CREATE_KV_INDICES(value_class, __VA_ARGS__)                                                            \
+                                                                                                                       \
+      table_class(eosio::name contract_name) {                                                                         \
+         init(contract_name, table_name##_n, db_name EOSIO_CDT_LIST_KV_INDICES(__VA_ARGS__));                          \
+      }                                                                                                                \
+   };
+
+/**
+ * @brief Macro to define an index.
+ * @details In the case where the autogenerated index names created by DEFINE_TABLE are not enough, a user can instead
+ * manually define the table and indices. This macro allows users to conveniently define an index without having to specify
+ * the index template type, as those can be large/unwieldy to type out.
+ *
+ * @param index_name    - The index name.
+ * @param member_name   - The name of the member pointer used for the index. This also defines the index's C++ variable name.
+ */
+#define KV_NAMED_INDEX(index_name, member_name)                                                                        \
+   index<EOSIO_CDT_GET_RETURN_T(value_type, member_name)> member_name{index_name ## _n, &value_type::member_name};
```

Comment:
Again use `eosio::name(index_name)` rather than `index_name ## _n`.

Also this assumes that the `value_type` type alias is defined in the table struct. For consistency the `DEFINE_TABLE` macro should define this type alias. Also the example documentation to include that and make a note about how it is expected that the `value_type` type alias is properly defined when manually constructing the table struct.

Reply:
The value_type alias is defined in the base table class, so we know it will always be there.

Reply:
Oh I see. Still working my way through the code.